Add support for "dialects"

nojb opened the pull request:
Following-up on #2373 and @diml's [comment](https://github.com/ocaml/dune/pull/2373#issuecomment-509156003), we propose the introduction of a `(dialect ...)` stanza. Full form is:

```
(dialect
  (name <name>)
  (interface
    (extension <string>)
    (preprocess <action>)
    (format <action>))
  (implementation
    (extension <string>)
    (preprocess <action>)
    (format <action>)))
```

- Fields `preprocess` and `format` are optional, in which case it is assumed that the files in question are already valid OCaml (and/or do not need special formatting).
- Dialects are uniquely identified by their name on a given project
- Each dialect has exactly two associated extensions: one for interfaces and one for implementations. The extensions are unique among all dialects of a project.

nojb committed:
Add Dialect module

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Add Dialect interface

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Include . in extension

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Cleanup

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented:
In the first few commits I added internal plumbing to reimplement the existing logic for OCaml/Reason (not yet the formatting part).

nojb commented on src/dialect.ml:
```diff
@@ -0,0 +1,106 @@
+    | Action (loc, action) ->
+        constr "action"
+          [ Loc.to_dyn loc
+          ; Dune_lang.to_dyn (Action_dune_lang.encode action)
```
Is this the right way to "dyn"-ify an action? By the way, what is the purpose of `Dyn`? Is it used for serialization?

diml replied:
That seems right. `dyn` is only used for crash error reports (`Code_error.raise`). It used in tests as well sometimes. Basically, it's a format aimed at Dune developers rather than the user.

nojb commented on src/dialect.ml:
```diff
@@ -0,0 +1,106 @@
+        ; S.virt     __POS__ "binary"
+        ; S.virt_var __POS__ "input-file"
+        ]
+    in
```
Implementing the preprocessing of Reason files in this way results in a worse UI (we don't get the hint: `try: opam install reason`).

diml replied:
For the hint, I suggest to add something in `Action.Prog.Not_found.raise`: if the `hint` is `None` and and the program name is `reason`, then we add this hint.

Apart for the hint, is the UI worse in other ways?

nojb replied:
No, that's the only thing, as far as I'm aware.

nojb committed:
Remove debug

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Add plumbing for format action

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

diml commented on src/dialect.mli:
```diff
@@ -0,0 +1,25 @@
+open! Stdune
```
Could you add a comment to describe what this module is about?

diml replied:
That's a good description for the manual BTW :)

diml commented on src/module.ml:
```diff
@@ -46,21 +36,14 @@ module Name = struct
-      | OCaml, Impl -> ".ml"
-    in
-    String.lowercase n ^ ext
+  let basename n ~(ml_kind : Ml_kind.t) ~(syntax : Dialect.t) =
```
We should rename the argument to `~dialect` to make things clearer

nojb replied:
Yes, done.

diml commented on src/module.ml:
```diff
@@ -46,21 +36,14 @@ module Name = struct
   type t =
     { path   : Path.t
-    ; syntax : Syntax.t
+    ; syntax : Dialect.t
```
Same, here should rename the field `dialect`

nojb replied:
Done.

diml commented on src/preprocessing.ml:
```diff
@@ -635,6 +616,24 @@ let action_for_pp sctx ~dep_kind ~loc ~expander ~action ~src ~target =
   | None -> action
   | Some dst -> Action.with_stdout_to (Path.build dst) action
 
+(* Generate rules for the reason modules in [modules] and return a
```
This comment should be generalised since we now support more than just the `reason` dialect

nojb replied:
Done.

diml commented:
I like the name `dialect` BTW :)

nojb committed:
Restore hint about installing reason if refmt is not found

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Add doc

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

emillon commented:
If I understand it correctly, the interface expected for `(format <action>)` is using stdin/stdout. Unfortunately, some formatters like `ocamlformat` don't support this. Do you think there's a way to support that?

nojb committed:
Rename 'syntax' to 'dialect'

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented:
> If I understand it correctly, the interface expected for `(format <action>)` is using stdin/stdout. Unfortunately, some formatters like `ocamlformat` don't support this. Do you think there's a way to support that?

- `dune` preprocessing actions write to `stdout` so it makes sense to do the same here
- if you are formatting ocaml code (ie. you are not specifying a `preprocess` action for the dialect) you do not need to actually write the formatting action and `dune` knows how to invoke `ocamlformat`
- why doesn't `ocamlformat` support writing on `stdout` ?? 😄

nojb committed:
Fix doc

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Add Dialect.S module

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

diml commented:
@emillon it's only printing to `stdout` that's expected. The input file is passed via a variable `%{input-file}`. For instance:

```
(format (run ocamlformat %{input-file}))
```

nojb commented:
Ah, sorry for the confusion, I thought the question was about printing to stdout rather than reading from stdin.

diml commented on src/dir_contents.ml:
```diff
@@ -97,10 +97,10 @@ let modules_of_files ~dir ~files =
+      | Some (s, "ml" ) -> Left  (make_module Dialect.ocaml  s fn)
+      | Some (s, "re" ) -> Left  (make_module Dialect.reason s fn)
+      | Some (s, "mli") -> Right (make_module Dialect.ocaml  s fn)
+      | Some (s, "rei") -> Right (make_module Dialect.reason s fn)
```
I guess we should look in an `extension -> dialect` dictionary here?

nojb replied:
Indeed, I am looking at doing that now.

nojb replied:
This is done now.

diml commented on src/format_rules.ml:
```diff
@@ -71,10 +71,15 @@ let gen_rules_output sctx (config : Dune_file.Auto_format.t) ~output_dir =
-      | _, ".rei" when Dune_file.Auto_format.includes config Reason ->
-        let refmt = Refmt.get sctx ~loc:(Some loc) ~dir in
-        Some (Refmt.format refmt ~input ~output)
+      | _, (".re"|".rei" as ext) when Dune_file.Auto_format.includes config Reason ->
```
And here as well?

nojb replied:
This is also done.

nojb committed:
Split Action.Make_mapper off and move Ml_kind.{flag,ppx_driver_flag} to break cycle

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Add dialects to Dune_project.t

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Module.modules_of_files: use project dialects

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Format_rules.gen_rules_output: use project dialects

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Extend (using fmt ...) to work with arbitrary dialects

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Add dune_lang decoder for dialect stanzas

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Record defined dialects in dune-project

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Fix

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Fix

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Use ocamlformat when possible

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
fmt enabled for all dialects by default in version 1.1

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Add test

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Only include all dialects in version > v1.1 of fmt

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Remove trailing semicolon

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Remove dead module Refmt

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Promote test result

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Tweak

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Fix (enabled_for ...) decoding

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Restrict (dialect ...) stanza to >= 1.11

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Bump fmt version

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Error handling

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Better error reporting

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Fix

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Elaborate ocamlformat rule from dialect definition

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Fail if extension contains a dot

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented:
This is ready for review.

nojb commented on src/dialect.ml:
```diff
@@ -0,0 +1,181 @@
+      Action_dune_lang.run (S.virt __POS__ "refmt")
+        [ S.virt     __POS__ "--print"
+        ; S.virt     __POS__ "binary"
+        ; S.virt_var __POS__ "input-file"
```
Using `__POS__` here is wrong because in case of error the location points to `dune`'s source code. I guess we want to pass `Loc.none` somehow.

rgrinberg replied:
If you don't want this, I think you can just use `make_var` and `make_text`. You can pass any location you want this way.

nojb replied:
Thanks, I did this change.

nojb committed:
Adapt ocamlformat test

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Promote test result

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Fix definition of anonymous dune-project

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented on src/dune_file.ml:
```diff
@@ -641,35 +641,28 @@ module Auto_format = struct
 
   let enabled_for_field =
-    let+ r = field_o "enabled_for" (repeat language)
+    let+ r = field_o "enabled_for" (repeat (map ~f:language string))
```
We do not validate that the language names used in the `(enabled_for ...)` field actually exist (i.e. are either `dune` or a dialect name). This validation has to happen after the `dune-project` file has been parsed, but the "extensions" API would need to be extended to allow doing such a thing.

nojb commented on src/dialect.mli:
```diff
@@ -0,0 +1,54 @@
+
+val preprocess : t -> Ml_kind.t -> (Loc.t * Action_dune_lang.t) option
+
+val format : t -> Ml_kind.t -> (Loc.t * Action_dune_lang.t * string list) option
```
The `string list` is a list of "extra dependencies" for the formatting rule. This is only used to record the dependency of the `ocamlformat` rule on `.ocamlformat` and `.ocamlformat-ignore`, and is not exposed to the user.

rgrinberg commented on src/dialect.ml:
```diff
@@ -0,0 +1,181 @@
+  let kind kind =
+    let+ loc, extension  = field "extension" (located string)
+    and+ preprocess = field_o "preprocess" (located Action_dune_lang.decode)
+    and+ format     = field_o "format" (map ~f:(fun (loc, x) -> loc, x, []) (located Action_dune_lang.decode))
```
Could you stick to an 80 char limit? It makes reading diffs much easier.

nojb replied:
Sure, it is fixed.

nojb committed:
Wrap long lines

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

rgrinberg committed:
Clean up formatting code with monads

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented on src/action_mapper.ml:
```diff
@@ -0,0 +1,47 @@
+module Make
+    (Src : Action_intf.Ast)
+    (Dst : Action_intf.Ast)
+= struct
```
Can we include an .mli for this module? I assume you had to split it because of circular deps?

nojb replied:
Yes, it was to break a dependency cycle. I refrained from adding an `mli` because both functions need to be exposed and they have a long type that needs to be repeated, and it is not convenient to add a type component to the signature in order to share this type because the module is `included`-ed in other modules that may define similarly-named types.

rgrinberg commented on test/blackbox-tests/test-cases/formatting/run.t:
```diff
@@ -87,7 +87,7 @@ And fixable files can be promoted:
   Promoting _build/default/enabled/.formatted/ocaml_file.ml to enabled/ocaml_file.ml.
   Promoting _build/default/enabled/.formatted/reason_file.re to enabled/reason_file.re.
   $ cat enabled/ocaml_file.ml
-  Sys.argv: ../install/default/bin/ocamlformat --impl enabled/ocaml_file.ml --name ../../enabled/ocaml_file.ml -o enabled/.formatted/ocaml_file.ml
```
How come these arguments aren't passed anymore?

nojb replied:
Now we write the output to stdout, so the `-o` flag is not necessary anymore. And instead of using `--name` (which takes a filename to be used in error messages), I changed the action so that it `chdir`s to `%{workspace_root}` which produces nicer paths in the error messages.

rgrinberg replied:
>  I changed the action so that it chdirs to %{workspace_root} which produces nicer paths in the error messages.

AFAIK that was already in place. But ok, I'm convinced the arguments aren't needed.

rgrinberg commented on src/format_rules.ml:
```diff
@@ -29,60 +24,58 @@ let depend_on_files ~named dir =
 let formatted = ".formatted"
 
-let gen_rules_output sctx (config : Dune_file.Auto_format.t) ~output_dir =
+let gen_rules_output sctx (config : Dune_file.Auto_format.t) ~dialects ~expander ~output_dir =
```
80 chars

rgrinberg commented on src/dialect.ml:
```diff
@@ -0,0 +1,191 @@
+
+  let find_by_extension { by_extension ; _ } extension =
+    Option.map ~f:(fun dialect ->
+      let kind =if dialect.file_kinds.intf.extension = extension then
```
why don't we check for the impl extension in the else clause?

nojb replied:
This map is constructed in a way (when calling `Dialect.S.add`) that if an extension is a key then it is the extension of either the interface or the implementation of the corresponding value.

rgrinberg reviewed (approved):
Looks good. Feel free to merge once my small nitpicks are addressed.

nojb committed:
Wrap long line

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Missing space

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented:
Thanks!

nojb committed:
Add manual doc

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented:
I added some doc for the manual. I didn't know where to put it so I added a new top-level section (called `Dialects`). But the functionality is pretty minor, so it would probably be better to insert it inside some existing section.

nojb committed:
Use Loc.none instead of __POS__

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Update CHANGES.md

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Tweak error message, add more tests

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented:
I added a few more tests. I will merge after the CI passes unless someone objects.

diml commented on src/dialect.mli:
```diff
@@ -0,0 +1,56 @@
+
+val ml_suffix : t -> Ml_kind.t -> string option
+
+module S : sig
```
I forgot to comment about this, but what do you think about calling this module `DB`? This is the name we used in other places for this kind of things:

- `Lib.DB`
- `Scope.DB`

nojb replied:
Good point. I will include this change (and anything else that comes up) in #2411

diml replied:
Thanks, I had a quick look through the code. Apart from the comment about `loc`, it all looks good to me!

diml commented on src/dialect.mli:
```diff
@@ -0,0 +1,56 @@
+  type t
+
+  val empty : t
+  val add : t -> ?loc:Loc.t -> dialect -> t
```
It's better to make `loc` required here. We can always pass `Loc.none` as a dummy value, but at least we have to think about what we are doing and whether it is justified.

nojb replied:
Yes, will do.

bikallem commented:
This is interesting. I am wondering if this could be used to implement a JSX syntax in ocaml without the use of ppx, i.e. JSX syntax which is supported as part of ocaml syntax (like in ReasonML). Here the scope of the dialect is just to convert JSX into function application and nothing more. 

More information can be found in https://github.com/ocaml/ocaml/issues/8802. 

How would one go about doing it using the dialect support in ocaml? Is there a simple example repo that I can study/look? 

Could the following be perhaps my starting point?

1. Fork just the parsing bit from https://github.com/ocaml/ocaml/tree/trunk/parsing.
2. Implement JSX syntax on top of it
3. Implement AST transformation from JSX to ocaml AST.

nojb commented:
> 1. Fork just the parsing bit from https://github.com/ocaml/ocaml/tree/trunk/parsing.
> 2. Implement JSX syntax on top of it
> 3. Implement AST transformation from JSX to ocaml AST.

That sounds like it should work; but maintaining your fork of the OCaml parser is probably going to be quite a bit of work.

bikallem commented:
> That sounds like it should work; but maintaining your fork of the OCaml parser is probably going to be quite a bit of work.

Yep, that is my main worry too. It would have been nice if menhir supported these type of scenarios, i.e. inheriting the ocaml compiler parser without modifying it and only extending certain sections of it in separate .mly(?) file. Much like in OO how child classes can further specialize parent function/methods without having to maintain/modify the parent classes. 

This would make the maintenance burden of small dialects like I am proposing much palatable. Any ideas suggestions if this is possible perhaps?

diml commented:
Note that this is exactly what Camlp4 allowed to do; it allowed to extend the syntax of OCaml without forking the parser. Camlp4 is the predecessor of ppx. One issue with this approach is that we ended up with many different dialects and it was impossible for editors and tools to know them all. That's why we introduced the extension point and attribute syntax: to provide a uniform syntax for extending the language.

Given that, I wouldn't recommend extending the syntax in this way. Instead, you could design a syntax based on plain extension points, or embed the foreign syntax in a string: `[%jsx{| <div> ... </div> |}]`.
